### ISSUE TITLE
regression: piping into @() broken after process control chagnes

### ISSUE THREAD

**Issue by con-f-use** [xonsh/xonsh#3368]
The recent process control changes (https://github.com/xonsh/xonsh/commit/5648b8d528824aaca3776d33578dc6a802e6d9ae, PR: https://github.com/xonsh/xonsh/pull/3342) have broken piping on some systems.

According to the [documentation]() the following code should work:

**Expected Behavior**
```
# using commit d782074
$ echo "hello" | @(lambda a, s=None: s.read().strip() + " world\n")            
hello world
```

**Observed Behavior**
```
# using commit 5648b8d
$ echo "hello" | @(lambda a, s=None: s.read().strip() + " world\n")            
^C^C^C^C^C^X   # Waiting forever, cannot abort with Ctrl + C
```
Seen on Ubuntu 16.04, 18.04, 19.10

See:
 - https://github.com/xonsh/xonsh/issues/3366

**Comment by laloch** [comment]
@con-f-use, thanks for reporting!
Not only `@()` is affected. Lydia branch seems to be very racy. Turns out that even very basic `ls -1 | sort` yields
```
$ xonsh --no-rc
[...]
$ $XONSH_SHOW_TRACEBACK = True
$ ls -1 | sort                                                                                                   <function prompt_builder.<locals>.prompt at 0x7f4a4ff3bcb0>
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/laloch/code/xonsh/xonsh/proc.py", line 1886, in __init__
    proc = spec.run(pipeline_group=pipeline_group)
  File "/home/laloch/code/xonsh/xonsh/built_ins.py", line 549, in run
    p = self._run_binary(kwargs)
  File "/home/laloch/code/xonsh/xonsh/built_ins.py", line 560, in _run_binary
    p = self.cls(self.cmd, bufsize=bufsize, **kwargs)
  File "/home/laloch/code/xonsh/xonsh/proc.py", line 558, in __init__
    self._disable_suspend_keybind()
  File "/home/laloch/code/xonsh/xonsh/proc.py", line 836, in _disable_suspend_keybind
    termios.tcsetattr(0, termios.TCSANOW, mode)
termios.error: (4, 'Interrupted system call')
```

**Comment by con-f-use** [comment]
:sob: and I was so happy backgrounding finally worked...

**Comment by con-f-use** [comment]
Btw. your basic `ls -1 | sort` works for me.

<details>

        confus@jcgbx ~ $ ls -1 | sort                                                😱  😱
        Desktop
        devel
        docfancypants.bash
        Documents
        Downloads
        Pictures
        Public
        Templates
        Videos
        confus@jcgbx ~ $ xonfig                                                      😱  😱
        +------------------+----------------------+
        | xonsh            | 0.9.13               |
        | Git SHA          | 5648b8d5             |
        | Commit Date      | Oct 17 19:02:06 2019 |
        | Python           | 3.6.8                |
        | PLY              | 3.11                 |
        | have readline    | True                 |
        | prompt toolkit   | 2.0.9                |
        | shell type       | prompt_toolkit2      |
        | pygments         | 2.2.0                |
        | on posix         | True                 |
        | on linux         | True                 |
        | distro           | ubuntu               |
        | on darwin        | False                |
        | on windows       | False                |
        | on cygwin        | False                |
        | on msys2         | False                |
        | is superuser     | False                |
        | default encoding | utf-8                |
        | xonsh encoding   | utf-8                |
        | encoding errors  | surrogateescape      |
        +------------------+----------------------+

</details>

**Comment by laloch** [comment]
grr, another one:
```
$ ls -1 | head
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/laloch/code/xonsh/xonsh/proc.py", line 1886, in __init__
    proc = spec.run(pipeline_group=pipeline_group)
  File "/home/laloch/code/xonsh/xonsh/built_ins.py", line 549, in run
    p = self._run_binary(kwargs)
  File "/home/laloch/code/xonsh/xonsh/built_ins.py", line 560, in _run_binary
    p = self.cls(self.cmd, bufsize=bufsize, **kwargs)
  File "/home/laloch/code/xonsh/xonsh/proc.py", line 558, in __init__
    self._disable_suspend_keybind()
  File "/home/laloch/code/xonsh/xonsh/proc.py", line 836, in _disable_suspend_keybind
    termios.tcsetattr(0, termios.TCSANOW, mode)
termios.error: (5, 'Input/output error')
```

**Comment by con-f-use** [comment]
I also noticed that `history show | grep pip` doesn't work anymore. Likely related.

**Comment by scopatz** [comment]
hmm this seems to work for me now

**Comment by laloch** [comment]
Yep, can't reproduce on v0.9.14 either.

**Comment by gforsyth** [comment]
Sorry, this was fixed by #3419 but I didn't realize we had another open issue around it.  I'm going to close this, but please ping if there are piping issues on 0.9.14 or later!
